File: the-lab-monorepo/apps/retailer-app-android/app/src/main/java/com/thelab/retailer/data/auth/FirebaseAuthHelper.kt
package com.thelab.retailer.data.auth

import android.content.Context
import android.util.Log
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.tasks.await

/**
 * Firebase Auth helper for dual-mode authentication.
 * Connects to Firebase Auth Emulator in debug builds.
 * All methods degrade gracefully — if Firebase is unavailable, legacy JWT still works.
 */
object FirebaseAuthHelper {
    private const val TAG = "FirebaseAuth"
    private var initialized = false

    /**
     * Initialize Firebase with programmatic config (no google-services.json needed for auth).
     * Call once from Application.onCreate().
     */
    fun init(context: Context) {
        if (initialized) return
        try {
            if (FirebaseApp.getApps(context).isEmpty()) {
                val options = FirebaseOptions.Builder()
                    .setProjectId("demo-thelab")
                    .setApplicationId("1:000000000000:android:0000000000000001")
                    .setApiKey("demo-key")
                    .build()
                FirebaseApp.initializeApp(context, options)
            }
            // Connect to emulator in debug builds
            if (com.thelab.retailer.BuildConfig.DEBUG) {
                val emulatorHost = "10.0.2.2" // Android emulator localhost
                FirebaseAuth.getInstance().useEmulator(emulatorHost, 9099)
            }
            initialized = true
            Log.d(TAG, "Firebase Auth initialized (debug=${com.thelab.retailer.BuildConfig.DEBUG})")
        } catch (e: Exception) {
            Log.w(TAG, "Firebase Auth init failed (non-fatal): ${e.message}")
        }
    }

    /**
     * Exchange a Firebase Custom Token from the backend for a Firebase session.
     * Returns the Firebase ID token string, or null on failure.
     */
    suspend fun exchangeCustomToken(customToken: String): String? {
        if (customToken.isBlank()) return null
        return try {
            val result = FirebaseAuth.getInstance().signInWithCustomToken(customToken).await()
            result.user?.getIdToken(false)?.await()?.token
        } catch (e: Exception) {
            Log.w(TAG, "Custom token exchange failed (non-fatal): ${e.message}")
            null
        }
    }

    /**
     * Get a fresh Firebase ID token for the currently signed-in user.
     * Returns null if no Firebase session exists.
     */
    suspend fun getIdToken(): String? {
        return try {
            FirebaseAuth.getInstance().currentUser
                ?.getIdToken(false)
                ?.await()
                ?.token
        } catch (e: Exception) {
            null
        }
    }

    /** Sign out of Firebase Auth. */
    fun signOut() {
        try {
            FirebaseAuth.getInstance().signOut()
        } catch (_: Exception) { }
    }
}

File: pegasus/apps/driverappios/driverappios/Services/FirebaseAuthHelper.swift
import Foundation

// MARK: - Firebase Auth Helper (Stub)
//
// Firebase Auth integration for dual-mode authentication.
// This file is a stub that stores the Firebase custom token for future
// exchange once the Firebase/Auth SPM package is added to the Xcode project.
//
// TODO: Add Firebase/Auth SPM dependency in Xcode:
//   1. File → Add Package Dependencies → https://github.com/firebase/firebase-ios-sdk
//   2. Select "FirebaseAuth" product
//   3. Uncomment the Firebase calls below
//
// Once FirebaseAuth is available, the flow is:
//   1. Backend returns firebase_token (custom token)
//   2. Call Auth.auth().signIn(withCustomToken:) to exchange it
//   3. Use auth.currentUser?.getIDToken() for API calls

final class FirebaseAuthHelper {
    static let shared = FirebaseAuthHelper()
    
    /// The Firebase ID token (after custom token exchange). Preferred over legacy JWT.
    private(set) var idToken: String?
    
    private init() {}
    
    /// Initialize Firebase for auth. Call once from app launch.
    /// In debug, connects to the emulator on localhost:9099.
    func initialize() {
        // TODO: Uncomment when FirebaseAuth SPM is added:
        // import FirebaseCore
        // import FirebaseAuth
        //
        // if FirebaseApp.app() == nil {
        //     let options = FirebaseOptions(googleAppID: "1:000000000000:ios:0000000000000000", gcmSenderID: "000000000000")
        //     options.projectID = "demo-thelab"
        //     options.apiKey = "demo-key"
        //     FirebaseApp.configure(options: options)
        // }
        // #if DEBUG
        // Auth.auth().useEmulator(withHost: "localhost", port: 9099)
        // #endif
    }
    
    /// Exchange a Firebase Custom Token from the backend for an ID token.
    /// Returns the ID token string via completion. Degrades gracefully.
    func exchangeCustomToken(_ customToken: String, completion: @escaping (String?) -> Void) {
        guard !customToken.isEmpty else {
            completion(nil)
            return
        }
        // TODO: Uncomment when FirebaseAuth SPM is added:
        // Auth.auth().signIn(withCustomToken: customToken) { [weak self] result, error in
        //     guard error == nil, let user = result?.user else {
        //         print("[FirebaseAuth] Custom token exchange failed: \(error?.localizedDescription ?? "unknown")")
        //         completion(nil)
        //         return
        //     }
        //     user.getIDToken { token, _ in
        //         self?.idToken = token
        //         completion(token)
        //     }
        // }
        
        // Stub: store the custom token as-is for now (will be replaced with ID token after SDK integration)
        print("[FirebaseAuth] Stub — custom token received, awaiting SDK integration")
        completion(nil)
    }
    
    /// Get a fresh Firebase ID token. Returns nil if no session exists.
    func getIdToken(completion: @escaping (String?) -> Void) {
        // TODO: Uncomment when FirebaseAuth SPM is added:
        // guard let user = Auth.auth().currentUser else {
        //     completion(nil)
        //     return
        // }
        // user.getIDToken { token, _ in completion(token) }
        completion(idToken)
    }
    
    /// Sign out of Firebase Auth.
    func signOut() {
        idToken = nil
        // TODO: Uncomment when FirebaseAuth SPM is added:
        // try? Auth.auth().signOut()
    }
}
